source/retryable-reads/tests/etc/templates/handshakeError.yml.template

```diff
@@ -0,0 +1,155 @@
+        - connectionCheckOutStartedEvent
+        - commandStartedEvent
+        - commandSucceededEvent
+        - commandFailedEvent
+  - database:
+      id: &database database
+      client: *client
+      databaseName: &databaseName retryable-reads-handshake-tests
+  - collection:
+      id: &collection collection
+      database: *database
+      collectionName: &collectionName coll
+
+initialData:
+  - collectionName: *collectionName
+    databaseName: *databaseName
+    documents:
+      - { _id: 1, x: 11 }
+      - { _id: 2, x: 22 }
+      - { _id: 3, x: 33 }
+
+tests:
+  # Because setting a failPoint creates a connection in the connection pool, run
+  # a ping operation that fails immediately after the failPoint operation in
+  # order to discard the connection before running the actual operation to be
+  # tested. The saslContinue command is used to avoid SDAM errors.
+  #
+  # Description of events:
+  # - Failpoint operation.
+  #   - Creates a connection in the connection pool that must be closed.
+  # - Ping operation.
+  #   - Triggers failpoint (first time).
+  #   - Closes the connection made by the fail point operation.
+  # - Test operation.
+  #   - New connection is created.
+  #   - Triggers failpoint (second time).
+  #   - Tests whether operation successfully retries the handshake and succeeds.
+{% for operation in operations %}
+  - description: "{{operation.operation_name}} succeeds after retryable handshake network error"
+    operations:
+      - name: failPoint
+        object: testRunner
+        arguments:
+          client: *client
+          failPoint:
+            configureFailPoint: failCommand
+            mode: { times: 2 }
+            data:
+              failCommands: [ping, saslContinue]
+              closeConnection: true
+      - name: runCommand
+        object: *database
+        arguments: { commandName: ping, command: { ping: 1 } }
+        expectError: { isError: true }
+      - name: {{operation.operation_name}}
+        object: *{{operation.object}}
+        {%- if operation.arguments|length > 0 %}
+        arguments:
+          {%- for arg in operation.arguments %}
+          {{arg}}
+          {%- endfor -%}
+        {%- endif %}
+    expectEvents:
+      - client: *client
+        eventType: cmap
+        events:
+          - { connectionCheckOutStartedEvent: {} }
+          - { connectionCheckOutStartedEvent: {} }
+          - { connectionCheckOutStartedEvent: {} }
+          - { connectionCheckOutStartedEvent: {} }
+      - client: *client
+        events:
+          - commandStartedEvent:
+              command: { ping: 1 }
+              databaseName: *databaseName
+          - commandFailedEvent:
+              commandName: ping
+          - commandStartedEvent:
+              commandName: {{operation.command_name}}
+          - commandSucceededEvent:
+              commandName: {{operation.command_name}}
+          {%- if operation.operation_name == "createChangeStream" %}
```

Comment:
There shouldn't be a killCursors here (and same for the second test):
```
FAIL: test_createChangeStream_succeeds_after_retryable_handshake_network_error (__main__.TestUnifiedHandshakeError)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/shane/git/mongo-python-driver/test/unified_format.py", line 1647, in test_case
    self.run_scenario(spec)
  File "/Users/shane/git/mongo-python-driver/test/unified_format.py", line 1600, in run_scenario
    self._run_scenario(spec, uri)
  File "/Users/shane/git/mongo-python-driver/test/unified_format.py", line 1630, in _run_scenario
    self.check_events(expect_events)
  File "/Users/shane/git/mongo-python-driver/test/unified_format.py", line 1554, in check_events
    self.assertEqual(len(actual_events), len(events), actual_events)
AssertionError: 4 != 6 Actual commands: [
<CommandStartedEvent ('localhost', 27017) db: 'retryable-reads-handshake-tests', command: 'ping', operation_id: 681910962, service_id: None>,
<CommandFailedEvent ('localhost', 27017) command: 'ping', operation_id: 681910962, duration_micros: 276, failure: {'errmsg': 'connection closed', 'errtype': 'OSError'}, service_id: None>,
<CommandStartedEvent ('localhost', 27017) db: 'retryable-reads-handshake-tests', command: 'aggregate', operation_id: 1260973671, service_id: None>,
<CommandSucceededEvent ('localhost', 27017) command: 'aggregate', operation_id: 1260973671, duration_micros: 4148, service_id: None>]
```

Since the aggregate failed there's no cursor to kill.

Reply:
Oh I see. The killCursors could come from the driver closing the cursor because it's not saved with saveResultAsEntity. In Python we cleanup these unsaved cursors at the very end of the test so they don't show up in the command events. Perhaps we can ignore killCursors events in these tests to avoid the potential inconsistency across drivers?

Reply:
I think we can use a `saveResultAsEntity` to ensure consistent `killCursors` behavior across Drivers.

Reply:
Good call.